File: UnitySokoban/Assets/Scripts/Planning/PlanGraphSGW/LiteralNode.cs
using Planning.Logic;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace PlanGraphSGW
{
    public class LiteralNode : Node
    {

        public Literal literal;
        internal List<StepNode> producers = new List<StepNode>();
        internal List<StepNode> consumers = new List<StepNode>();

        internal LiteralNode(PlanGraph graph, Literal literal)
            : base(graph)
        {
            this.literal = literal;
        }

        public override int GetHashCode()
        {
            return literal.GetHashCode();
        }
        public override string ToString()
        {
            return literal.ToString() + " " + level;
        }

        public override bool setLevel(int level)
        {
            if (base.setLevel(level))
            {
                foreach (StepNode consumer in consumers)
                    consumer.incrementLiteralCount();
                return true;
            }
            else
                return false;
        }

        public IEnumerable<StepNode> getProducers(int level)
        {
            IEnumerator<StepNode> iterator = new NodeIterator<StepNode>(level, producers);
            while (iterator.MoveNext())
                yield return iterator.Current;
        }

        public IEnumerable<StepNode> getConsumers(int level)
        {
            IEnumerator<StepNode> iterator = new NodeIterator<StepNode>(level, consumers);
            while (iterator.MoveNext())
                yield return iterator.Current;
        }
    }
}

File: Plan Graph (sgware)/src/edu/uno/ai/planning/pg/LiteralNode.java
package edu.uno.ai.planning.pg;

import java.util.ArrayList;
import java.util.Iterator;

import edu.uno.ai.planning.logic.Literal;

public class LiteralNode extends Node {

	public final Literal literal;
	protected final ArrayList<StepNode> producers = new ArrayList<>();
	protected final ArrayList<StepNode> consumers = new ArrayList<>();

	protected LiteralNode(PlanGraph graph, Literal literal) {
		super(graph);
		this.literal = literal;
	}
	
	@Override
	public int hashCode() {
		return literal.hashCode();
	}
	
	@Override
	public String toString() {
		return literal.toString();
	}
	
	@Override
	protected boolean setLevel(int level) {
		if(super.setLevel(level)) {
			for(StepNode consumer : consumers)
				consumer.incrementLiteralCount();
			return true;
		}
		else
			return false;
	}
	
	public Iterable<StepNode> getProducers(int level) {
		return new Iterable<StepNode>() {
			@Override
			public Iterator<StepNode> iterator() {
				return new NodeIterator<>(level, producers);
			}
		};
	}
	
	public Iterable<StepNode> getConsumers(int level) {
		return new Iterable<StepNode>() {
			@Override
			public Iterator<StepNode> iterator() {
				return new NodeIterator<>(level + 1, consumers);
			}
		};
	}
}
